components/brave_ads/browser/ads_notification_handler.cc

```diff
@@ -120,4 +157,47 @@ const void* AdsNotificationHandler::UserDataKey() {
   return &kUserDataKey;
 }
 
+void AdsNotificationHandler::OperationCompleted(
+    const std::string& notification_id) {
+
+  auto iter = pending_callbacks_.find(notification_id);
+  std::move(iter->second).Run();
+  pending_callbacks_.erase(iter);
+#if defined(OS_ANDROID)
+  StartShutDownTimerIfNecessary(notification_id);
+#endif
+}
+
+#if defined(OS_ANDROID)
+bool AdsNotificationHandler::IsHeadless(){
+  auto state = base::android::ApplicationStatusListener::GetState();
+  bool uiless = (state ==
+      base::android::ApplicationState::APPLICATION_STATE_UNKNOWN ||
+      state ==
+      base::android::ApplicationState::APPLICATION_STATE_HAS_DESTROYED_ACTIVITIES);
+  return uiless;
+}
+
+void AdsNotificationHandler::StartShutDownTimerIfNecessary (
+    const std::string & last_processed_notification_id){
+  if (IsHeadless() && last_dismissed_notification_id_ ==
+      last_processed_notification_id){
+    last_dismissed_notification_id_ = "";
+    //wait and close the browser if running headless
+    no_ui_shutdown_timer_.Stop();
+    no_ui_shutdown_timer_.Start(FROM_HERE,
+        base::TimeDelta::FromSeconds(kWaitBeforeShutdownWhenRunHeadless),
+        this, &AdsNotificationHandler::ShutdownTimerCallback);
+  }
+}
+
+//Check if browser is runnning without UI and close it.
+//If during shutdown a new notification event arrives-
+//timer is stopped in corresponding handler function
+void AdsNotificationHandler::ShutdownTimerCallback(){
+  if (IsHeadless()) {
```

Comment:
Is it safe to assume that any time Android is running in a headless mode, it's due to our ad notifications and is therefore safe to terminate? It seems like it would cause problems if that turned out not to be the case. At a minimum, we should probably have some kind of log here I think.

It seems like the issue we're trying to deal with is that the browser starts up with no UI when a notification is dismissed. Do we understand why that happens though? Maybe there's a less agressive way to deal with that scenario? cc: @bridiver

Reply:
I noticed, the Chromium 80 behaves the same way when a web notification is dismissed. But Brave starts serving ads in the background, which is not expected. We also don't want to consume an extra battery resources when it's not needed.

Reply:
If we don't serve ads when we're running in a headless state, would it be a decent compromise? Seems like it may be good to try to follow the "standard" Chromium behavior (such as it is) as much as possible, just to future-proof ourselves when they inevitably change something here.

Reply:
How about to save some battery charge?

Reply:
I don't know too much about the battery cost, but it seems that if Google allows this mode for web notifications they probably decided it was a negligible cost to pay.

Anyway, will approve on my side - don't want to hold this up, but curious what others think.